// i/config/orm.js
exports.orms = [
{
    name: 'User',
    abb: 'u',               // abbreviation for redis hash name
    child: null,
    column: [               // column to create bo file, column index is hash field in redis
        'userId',
        'level',
        'exp',
        'money',
        'energy',
        'energyLimit',
        'lastEnergyChargedTime',
        'lastLoginTime',
        'honor',
        'dailyVisitedFriends',
        'lastResetDailyVisitedFriendsTime',
        'side',
        'kingHeartLevel',
        'kingHeartAppearance',
        'loginCount',
    ],
    updateFilter: [0],
    clientFilter: [6, 7, 10, 14],
    pk: 'userId',               // primary key
    pkAutoIncrement: true,
    list: null,
},
{
    name: 'Card',
    abb: 'c',               // abbreviation for redis hash name
    child: null,
    column: [               // column to create bo file, column index is hash field in redis
        'cardId',
        'cardTypeId',
        'level',
        'exp',
        'strengthenLevel',
        'strengthenExp',
        'hp',
        'atk',
        'def',
        'hit',
        'dodge',
        'critical',
        'tough',
        'initAnger',
        'quality',
        'flag',
        'status',
        'todoStatus',
        'gemSlot',
    ],
    updateFilter: [0],
    clientFilter: [6, 7, 8, 9, 10, 11, 12, 13, 15, 16, 17],
    pk: 'cardId',               // primary key
    pkAutoIncrement: true,
    list: 'CardList',
},
];
