// i/lib/List.js
var List = function(listPK, list) {
    this.pk = listPK;
    this.reset(list);
};

List.prototype.reset = function(list) {
    this.list = list || {};

    this.toAddList = [];
    this.toDelList = [];
    this.toUpdateList = [];
};

List.prototype.add = function(child) {
    this.toAddList.push(child);
};

// input can be index or object
List.prototype.del = function(input) {
    var index = I.Util.isUInt(input) ? input : input.getPK();
    var child = this.get(index);
    this.toDelList.push(child);
};

List.prototype.update = function(child) {
    var index = child.getPK();
    // verify child exists
    this.get(index);

    this.toUpdateList.push(child);
};


List.prototype.get = function(pk) {
    var child = this.list[pk];
    if (!child) throw new I.Exception(10111);

    return child;
};

List.prototype.insert = function(child) {
    this.list[child.getPK()] = child;
};

List.prototype.remove = function(index) {
    delete this.list[index];
};

List.prototype.getKeys = function() {
    var keys = [];
    for (var i in this.list) {
        keys.push(i);
    }
    return keys;
};

List.prototype.getPK = function() { return this.pk; };

List.prototype.toClient = function() {
    var toClient = {};
    for (var i in this.list) {
        toClient[i] = this.list[i].toClient();
    }
    return toClient;
};

exports.List = List;
