// i/lib/Util.js
var Util = function Util() {};

/* Array & Object */
Util.prototype.getLength = function getLength(object) {
    if (Array.isArray(object)) {
        return object.length;
    } else if (typeof object === 'object') {
        var i = 0;
        for (var h in object) {
            ++i;
        }
        return i;
    } else {
        throw new I.Exception(10201);
    }
};

Util.prototype.getLastElement = function getLastElement(object) {
    if (Array.isArray(object)) {
        return object[object.length - 1];
    } else if (typeof object === 'object') {
        var length = this.getLength(object);
        var i = 0;
        for (var h in object) {
            if (i === length - 1) {
                return object[h];
            }
            ++i;
        }
    } else {
        throw new I.Exception(10202);
    }
};

Util.prototype.valueExist = function valueExist(value, object) {
    if (Array.isArray(object)) {
        return object.indexOf(value) !== -1;
    } else if (typeof object === 'object') {
        for (var h in object) {
            if (object[h] === value) {
                return true;
            }
        }
        return false;
    } else {
        throw new I.Exception(10206);
    }
};

/* Data Validation */
Util.prototype.isUInt = function isUInt(val) {
    var mid = parseInt(val);
    return mid == val && mid >= 0;
};
Util.prototype.isInt = function isInt(val) {
    return parseInt(val) == val;
};
Util.prototype.isArray = function isArray(array) {
    return Array.isArray(array);
};
Util.prototype.isHash = function isHash(hash) {
    return typeof hash === 'object';
};
Util.prototype.isString = function isString(string) {
    return typeof string === 'string';
};
Util.prototype.isEmpty = function isEmpty(val) {
    return val === '';
};

/* Date & Time */
Util.prototype.getTimestamp = function getTimestamp(time) {
    var stamp = this.getMicroTimestamp(time);
    return parseInt(stamp / 1000);
};

Util.prototype.getMicroTimestamp = function getMicroTimestamp(time) {
    if (time) {
        var stamp = Date.parse(time);
        if (!stamp) throw new I.Exception(10251);
        return stamp;
    }

    return Date.now();
};

/* Error */
Util.prototype.printError = function printError(error) {
    console.log(error.message + ':', I.ExceptionCodes[error.message]);
};

/* Probability */
Util.prototype.random = function random(start, end) {
    var n = end - start;
    return start + Math.floor(Math.random() * n);
};
Util.prototype.getElementByProbability = function getElementByProbability(probabilityList) {
    // verify probabilityList is hash
    if (!this.isHash(probabilityList)) throw new I.Exception(10203);

    // delete probability = 0 element & sum probabilities
    var all = 0;
    for (var i in probabilityList) {
        if (probabilityList[i] == 0) {
            delete probabilityList[i];
        } else {
            all += probabilityList[i];
        }
    }

    // no element in list
    if (all === 0) throw new I.Exception(10204);

    // get id by probability
    var seed = this.random(1, all);
    var sum = 0;
    for (var i in probabilityList) {
        sum += probabilityList[i];
        if (seed <= sum) {
            return i;
        }
    }

    throw new I.Exception(10205);
};

exports.Util = new Util();
